### PR TITLE
Feature/#440 hydra token login

### PR DESCRIPTION
**Pull request by Sheikah45**
This is a rework of #591 using the new hydra user service

### CODE REVIEW

File: server/oauth_service.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+from server.config import config
+from .core import Service
+from .decorators import with_logger
+from .exceptions import AuthenticationError
+
+
+@with_logger
+class OauthService(Service):
+    """
+        Service for managing the OAuth token logins and verification.
```
Comment (Askaholic) [inline server/oauth_service.py:15]: I decided I like this way better.
```suggestion
    Service for managing the OAuth token logins and verification.
```

File: server/oauth_service.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        # Run every day to update public keys.
+        self._update_cron = aiocron.crontab(
+            "0 0 * * *", func=self.retrieve_public_keys
+        )
+
+    async def retrieve_public_keys(self) -> None:
+        """
+            Get the latest jwks from the hydra endpoint
+        """
+        self.jwks_client = PyJWKClient(config.HYDRA_JWKS_URI)
```
Comment (Askaholic) [inline server/oauth_service.py:33]: Good news and bad news:
Bad news - this uses urllib which is blocking and pyjwt does not have any features for async
Good news - this class is pretty simple so it shouldn't be a big deal to write our own async version of it, or maybe just have the OAuthService do that directly
https://github.com/jpadilla/pyjwt/blob/69d1e8b5f450f2b1df42149d977b48e286df1054/jwt/jwks_client.py

File: server/oauth_service.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    async def get_player_id_from_token(self, token: str) -> int:
+        """
+            Decode the JWT to get the player_id
+        """
+        try:
+            signing_key = self.jwks_client.get_signing_key_from_jwt(token)
+            decoded_token = jwt.decode(token, signing_key.key, algorithms=["RS256"])
+            self._logger.info("Decoded Token: %s", decoded_token)
```
Comment (Askaholic) [inline server/oauth_service.py:42]: Not sure we want these in the prod logs. Probably `debug` is better.

File: tests/integration_tests/conftest.py [hunk 4]
```diff
@@ -10,7 +11,7 @@
 from aiohttp import web
 from asynctest import exhaust_callbacks
 
-from server import GameService, ServerInstance
+from server import GameService, ServerInstance, run_control_server
```
Comment (Askaholic) [inline tests/integration_tests/conftest.py:14]: Unused import

File: tests/data/test_conf.yaml [hunk 5]
```diff
@@ -19,6 +19,7 @@ API_CLIENT_SECRET: "banana"
 API_TOKEN_URI: "https://api.test.faforever.com/oauth/token"
 API_BASE_URL: "https://api.test.faforever.com/"
 USE_API: true
+API_JWT_PUBLIC_KEY_FILE: "tests/data/test-api-key.pub"
```
Comment (Askaholic) [inline tests/data/test_conf.yaml:22]: Don't think we need this anymore.

File: tests/integration_tests/test_login.py [hunk 6]
```diff
@@ -221,3 +224,213 @@ async def test_server_double_login(lobby_server):
...
+        "friends": [],
+        "foes": [],
+        "power": 0
+    }
+
+
+async def test_server_login_expired_token(lobby_server, api_priv_key):
+    proto = await connect_client(lobby_server)
+    await proto.send_message({
+        'command': 'auth',
```
Comment (Askaholic) [inline tests/integration_tests/test_login.py:295]: Oof  so many single quotes. Guess the base PR must be pretty dang old

File: tests/integration_tests/test_login.py [hunk 7]
```diff
@@ -221,3 +224,213 @@ async def test_server_double_login(lobby_server):
...
+        "command": "social",
+        "autojoin": ["#123_clan"],
+        "channels": ["#123_clan"],
+        "friends": [],
+        "foes": [],
+        "power": 0
+    }
+
+
+async def test_server_login_expired_token(lobby_server, api_priv_key):
```
Comment (Askaholic) [inline tests/integration_tests/test_login.py:397]: Tests are duplicated

**Review by Askaholic — COMMENTED** [review]
Need a test for the pub key refresh?

File: server/oauth_service.py (hunk 3 above)
Comment (Sheikah45) [reply server/oauth_service.py:42]: Yes I intend to remove this

File: server/oauth_service.py (hunk 2 above)
Comment (Sheikah45) [reply server/oauth_service.py:33]: Hmm okay then I will just use aiohttp then probably

File: tests/integration_tests/test_login.py (hunk 6 above)
Comment (Sheikah45) [reply tests/integration_tests/test_login.py]: Done

File: tests/data/test_conf.yaml (hunk 5 above)
Comment (Sheikah45) [reply tests/data/test_conf.yaml]: Done

File: tests/integration_tests/conftest.py (hunk 4 above)
Comment (Sheikah45) [reply tests/integration_tests/conftest.py]: Done

File: server/oauth_service.py (hunk 1 above)
Comment (Sheikah45) [reply server/oauth_service.py]: Done

File: tests/integration_tests/test_login.py (hunk 7 above)
Comment (Sheikah45) [reply tests/integration_tests/test_login.py]: Deduplicated

**Comment by Sheikah45** [comment]
This is now working and has decent tests. Further progress on some of the small details will need feedback from @Brutus5000 on how the Oauth will run on prod.

File: tests/data/test_conf.yaml [hunk 8]
```diff
@@ -47,4 +47,4 @@ LADDER_SEARCH_EXPANSION_MAX: 0.25
 LADDER_SEARCH_EXPANSION_STEP: 0.05
 QUEUE_POP_TIME_MAX: 180
 QUEUE_POP_DESIRED_MATCHES: 4
-QUEUE_POP_TIME_MOVING_AVG_SIZE: 5
+QUEUE_POP_TIME_MOVING_AVG_SIZE: 5
```
Comment (Askaholic) [inline tests/data/test_conf.yaml:50]: No pls

File: tests/integration_tests/conftest.py [hunk 9]
```diff
@@ -82,11 +98,15 @@ async def lobby_server(
...
-    policy_server
+    oauth_service,
+    policy_server,
+    jwks_server
 ):
     with mock.patch(
         "server.lobbyconnection.config.FAF_POLICY_SERVER_BASE_URL",
         f"http://{policy_server.host}:{policy_server.port}"
+    ), mock.patch("server.oauth_service.config.HYDRA_JWKS_URI",
+                        f"http://{jwks_server.host}:{jwks_server.port}/jwks"
```
Comment (Askaholic) [inline tests/integration_tests/conftest.py:109]: Formatting is a little janky

File: tests/integration_tests/conftest.py [hunk 10]
```diff
@@ -176,6 +196,53 @@ async def token(request):
...
+    app = web.Application()
+    routes = web.RouteTableDef()
+
+    class Handle(object):
+        def __init__(self):
+            self.host = host
+            self.port = port
+            self.result = json.loads(textwrap.dedent("""
+            {"keys":[
+            {
```
Comment (Askaholic) [inline tests/integration_tests/conftest.py:213]: No need to call `loads` here as JSON is also syntactically valid python.
```suggestion
            self.result = {
                "keys": [{
```

File: Pipfile [hunk 11]
```diff
@@ -20,11 +20,12 @@ trueskill = "*"
 aiocron = "*"
 oauthlib = "*"
 sqlalchemy = "*"
-twilio = "*"
+twilio = ">=6.0.0,<6.51.0"  # See https://github.com/twilio/twilio-python/issues/556
 humanize = ">=2.6.0"
 aiomysql = {editable = true, git = "https://github.com/aio-libs/aiomysql"}
 pyyaml = "*"
 aio_pika = "*"
+pyjwt = {version = "*", extras = ["crypto"]}
```
[the remark this answers is not in the document]
Comment (Sheikah45) [reply Pipfile:28]: Yep probably a good idea

File: server/oauth_service.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+                    self.public_keys[kid] = RSAAlgorithm.from_jwk(json.dumps(jwk))
+
+    async def get_player_id_from_token(self, token: str) -> int:
+        """
+            Decode the JWT to get the player_id
+        """
+        try:
+            kid = jwt.get_unverified_header(token)['kid']
+            key = self.public_keys[kid]
+            return int(jwt.decode(token, key=key, algorithms="RS256", options={"verify_aud": False})["sub"])
```
[the remark this answers is not in the document]
Comment (Sheikah45) [reply server/oauth_service.py:52]: Yes it stored as a string under sub. Potentially something Brutus can change in the token configuration

File: tests/data/test_conf.yaml (hunk 8 above)
Comment (Sheikah45) [reply tests/data/test_conf.yaml:50]: I am confused what changed here I will just change it back

Comment (Askaholic) [reply tests/data/test_conf.yaml:50]: You deleted the trailing newline

File: server/lobbyconnection.py [hunk 13]
```diff
@@ -505,14 +515,69 @@ async def check_policy_conformity(self, player_id, uid_hash, session, ignore_res
...
+            "command": "irc_password",
+            "password": new_irc_password
+        })
+
+        async with self._db.acquire() as conn:
+            result = await conn.execute(
+                select([t_login.c.login, t_login.c.steamid])
+                .where(t_login.c.id == player_id)
+            )
+            row = await result.fetchone()
```
Comment (Askaholic) [inline server/lobbyconnection.py:534]: Maybe we should have some handling for if the player id is not found. At least we should log a warning, and probably raise an authentication error with a useful message.

File: server/oauth_service.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+    def __init__(self):
+        self.public_keys = None
+
+    async def initialize(self) -> None:
+        await self.retrieve_public_keys()
+        # crontab: min hour day month day_of_week
+        # Run every day to update public keys.
+        self._update_cron = aiocron.crontab(
+            "0 0 * * *", func=self.retrieve_public_keys
```
Comment (Askaholic) [inline server/oauth_service.py:28]: It would probably be OK to run more frequently. Maybe every 10 minutes? Or implement some logic for detecting when token signatures no longer match.

File: server/oauth_service.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        )
+
+    async def retrieve_public_keys(self) -> None:
+        """
+        Get the latest jwks from the hydra endpoint
+        """
+        self.public_keys = {}
+        async with aiohttp.ClientSession(raise_for_status=True) as session:
+            async with session.get(config.HYDRA_JWKS_URI) as resp:
+                jwks = await resp.json()
```
Comment (Askaholic) [inline server/oauth_service.py:38]: I would only clear the dictionary once all coroutines have been awaited, otherwise you could have a situation where the dictionary is cleared, then while the get request is happening someone tries to log in and can't because there are no keys.

You could also use `dict.clear()` to reuse the same dictionary object instead of creating a new one each time. I don't think anything bad can happen by doing that because of the GIL.

```suggestion
        async with aiohttp.ClientSession(raise_for_status=True) as session:
            async with session.get(config.HYDRA_JWKS_URI) as resp:
                jwks = await resp.json()
                self.public_keys = {}
```

File: tests/integration_tests/conftest.py [hunk 16]
```diff
@@ -82,12 +98,14 @@ async def lobby_server(
...
-    policy_server
+    oauth_service,
+    policy_server,
+    jwks_server
 ):
-    with mock.patch(
-        "server.lobbyconnection.config.FAF_POLICY_SERVER_BASE_URL",
-        f"http://{policy_server.host}:{policy_server.port}"
-    ):
+    with mock.patch("server.lobbyconnection.config.FAF_POLICY_SERVER_BASE_URL",
```
Comment (Askaholic) [inline tests/integration_tests/conftest.py:105]: Oh I see what's going on here, two context managers in the same with statement. I didn't read that correctly last time. Using `\` line continuation is definitely a red flag for some possibly bad formatting. Maybe it would be better to assign the context managers to variables:
```python3
patch_policy = mock.patch(...)
patch_jwks = mock.patch(...)

with patch_policy, patch_jwks:
    ...
```

File: tests/integration_tests/test_login.py [hunk 17]
```diff
@@ -221,3 +224,109 @@ async def test_server_double_login(lobby_server):
...
+        "token": jwt.encode({
+          "sub": 3,
+          "user_name": "Rhiza",
+          "scope": [],
+          "exp": int(time() + 1000),
+          "authorities": [],
+          "non_locked": True,
+          "jti": "",
+          "client_id": ""
+        }, jwk_priv_key, algorithm="RS256", headers={"kid": "L7wdUtrDssMTb57A_TNAI79DQCdp0T2-KUrSUoDJBhk"}),
```
Comment (Askaholic) [inline tests/integration_tests/test_login.py:244]: Maybe it makes sense to define the jwk as a fixture as well, then we don't need to hardcode the kid here

File: tests/integration_tests/test_login.py [hunk 18]
```diff
@@ -221,3 +224,109 @@ async def test_server_double_login(lobby_server):
...
+
+    msg = await proto.read_message()
+    assert msg == {
+        "command": "authentication_failed",
+        "text": "Token signature was invalid"
+    }
+
+
+async def test_server_login_malformed_token(lobby_server, jwk_priv_key):
+    """This scenario could only happen if the API somehow signed a token that
```
Comment (Askaholic) [inline tests/integration_tests/test_login.py:314]: Maybe a test with a similar note for a token with a player id that doesn't exist in the database.

File: server/lobbyconnection.py [hunk 19]
```diff
@@ -505,14 +515,69 @@ async def check_policy_conformity(self, player_id, uid_hash, session, ignore_res
...
+    async def command_auth(self, message):
+        token = message["token"]
+        unique_id = message["unique_id"]
+        player_id = await self.oauth_service.get_player_id_from_token(token)
+
+        new_irc_password = hexlify(os.urandom(16)).decode()
+        await self.send({
+            "command": "irc_password",
+            "password": new_irc_password
+        })
```
Comment (Askaholic) [inline server/lobbyconnection.py:527]: Crazy thought. @Brutus5000 can we make the IRC server use hydra as well?

File: server/lobbyconnection.py [hunk 20]
```diff
@@ -505,14 +515,69 @@ async def check_policy_conformity(self, player_id, uid_hash, session, ignore_res
...
+                self.player_service.is_uniqueid_exempt(player_id)
+            )
+        )
+        if not conforms_policy:
+            return
 
+        self._logger.debug(
+            "Login from: %s, %s, %s", player_id, username, self.session
+        )
+        metrics.user_logins.labels("success").inc()
```
Comment (Askaholic) [inline server/lobbyconnection.py:578]: Do we want to include the type of login (password or token) in the metrics at all?

File: tests/integration_tests/test_login.py (hunk 17 above)
Comment (Sheikah45) [reply tests/integration_tests/test_login.py:244]: You mean the header for the jwk? As each test has a slightly different jwk.

File: server/oauth_service.py (hunk 14 above)
Comment (Sheikah45) [reply server/oauth_service.py:28]: Done

File: server/oauth_service.py (hunk 15 above)
Comment (Sheikah45) [reply server/oauth_service.py:38]: I made it so that it should create a new dictionary and then replaces the old dictionary after it is populated so that there should be practically no time with an empty public_keys

File: tests/integration_tests/test_login.py (hunk 18 above)
Comment (Sheikah45) [reply tests/integration_tests/test_login.py:314]: Done

File: server/lobbyconnection.py (hunk 20 above)
Comment (Sheikah45) [reply server/lobbyconnection.py:578]: Probably would be good

File: tests/integration_tests/conftest.py (hunk 16 above)
Comment (Sheikah45) [reply tests/integration_tests/conftest.py]: Done

File: server/lobbyconnection.py [hunk 21]
```diff
@@ -505,14 +516,74 @@ async def check_policy_conformity(self, player_id, uid_hash, session, ignore_res
...
+            ignore_result=(
+                steamid is not None or
+                self.player_service.is_uniqueid_exempt(player_id)
+            )
+        )
+        if not conforms_policy:
+            return
+
+        self._logger.debug(
+            "Login from: %s, %s, %s", player_id, username, self.session
```
Comment (Askaholic) [inline server/lobbyconnection.py:582]: Could add the method to the log statement too

File: tests/integration_tests/conftest.py [hunk 22]
```diff
@@ -176,6 +199,52 @@ async def token(request):
...
+    class Handle(object):
+        def __init__(self):
+            self.host = host
+            self.port = port
+            self.result = {
+                "keys": [{
+                    "kty": "RSA",
+                    "e": "AQAB",
+                    "use": "sig",
+                    "kid": "L7wdUtrDssMTb57A_TNAI79DQCdp0T2-KUrSUoDJBhk",
```
Comment (Askaholic) [inline tests/integration_tests/conftest.py:219]: Use the fixture here too. You can merge dictionaries like this:
```python3
>>> header = {"kid": "whatever"}
>>> {"foo": "bar", **header}
{'foo': 'bar', 'kid': 'whatever'}
```
But maybe it makes more sense to have the fixture just return the kid and format it appropriately wherever it's used.

File: tests/unit_tests/test_configuration_refresh.py [hunk 23]
```diff
@@ -54,17 +53,6 @@ async def test_config_refresh_empty_file(monkeypatch):
     assert config.DB_PASSWORD == "banana"
 
 
-@fast_forward(20)
-async def test_configuration_refresh(config_service, monkeypatch):
```
Comment (Askaholic) [inline tests/unit_tests/test_configuration_refresh.py:58]: Why is this deleted?

File: tests/integration_tests/test_login.py (hunk 17 above)
Comment (Askaholic) [reply tests/integration_tests/test_login.py]: I mean the kid really. The giant string that's the same in every test.

File: server/lobbyconnection.py [hunk 24]
```diff
@@ -153,6 +166,7 @@ async def on_message_received(self, message):
             await handler(message)
 
         except AuthenticationError as ex:
+            metrics.user_logins.labels(f"failure {ex.method}").inc()
```
[the remark this answers is not in the document]
Comment (Sheikah45) [reply server/lobbyconnection.py:169]: How would differ the labels beyond specifying the method?

File: server/lobbyconnection.py [hunk 25]
```diff
@@ -505,14 +516,74 @@ async def check_policy_conformity(self, player_id, uid_hash, session, ignore_res
...
+
+        async with self._db.acquire() as conn:
+            result = await conn.execute(
+                select([t_login.c.login, t_login.c.steamid])
+                .where(t_login.c.id == player_id)
+            )
+            row = await result.fetchone()
+
+            if not row:
+                raise AuthenticationError("User id not in database", auth_method)
```
[the remark this answers is not in the document]
Comment (Sheikah45) [reply server/lobbyconnection.py:533]: Ah gotcha for some reason I thought the error message was auto logged as well

File: server/oauth_service.py [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+        Get the latest jwks from the hydra endpoint
+        """
+        async with aiohttp.ClientSession(raise_for_status=True) as session:
+            async with session.get(config.HYDRA_JWKS_URI) as resp:
+                jwks = await resp.json()
+                new_keys = {}
+                for jwk in jwks["keys"]:
+                    kid = jwk["kid"]
+                    new_keys[kid] = RSAAlgorithm.from_jwk(jwk)
+                self.public_keys = new_keys
```
[the remark this answers is not in the document]
Comment (Sheikah45) [reply server/oauth_service.py:42]: Spent too much time in java forgot about the good ol comprehensions

File: tests/unit_tests/test_configuration_refresh.py (hunk 23 above)
Comment (Sheikah45) [reply tests/unit_tests/test_configuration_refresh.py:58]: Hmm I think I had added a different configuration and then removed it and this was accidental I will check.

File: server/lobbyconnection.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/lobbyconnection.py:169]: Look at `user_connections` for an example, it uses 2 labels

[the remark this answers is not in the document]
Comment (Sheikah45) [reply server/lobbyconnection.py:169]: Ah yeah you mean a whole separate type of label that can be done

File: server/lobbyconnection.py (hunk 21 above)
Comment (Sheikah45) [reply server/lobbyconnection.py:582]: Done

File: server/oauth_service.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (Sheikah45) [reply server/oauth_service.py:42]: Done

File: tests/integration_tests/conftest.py (hunk 22 above)
Comment (Sheikah45) [reply tests/integration_tests/conftest.py:219]: Makes sense

File: server/lobbyconnection.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (Sheikah45) [reply server/lobbyconnection.py:533]: Done

File: server/lobbyconnection.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (Sheikah45) [reply server/lobbyconnection.py]: Done

File: server/lobbyconnection.py [hunk 27]
```diff
@@ -505,14 +516,75 @@ async def check_policy_conformity(self, player_id, uid_hash, session, ignore_res
...
+
+        async with self._db.acquire() as conn:
+            result = await conn.execute(
+                select([t_login.c.login, t_login.c.steamid])
+                .where(t_login.c.id == player_id)
+            )
+            row = await result.fetchone()
+
+            if not row:
+                self._logger.error("User id from token not found in database possible fraudulent token")
```
Comment (Askaholic) [inline server/lobbyconnection.py:533]: But make sure it fits on 1 line
```suggestion
                self._logger.warning("User id from token not found in database possible fraudulent token: %s", player_id)
```

File: tests/integration_tests/conftest.py [hunk 28]
```diff
@@ -82,12 +103,15 @@ async def lobby_server(
...
+    oauth_service,
+    policy_server,
+    jwks_server
 ):
-    with mock.patch(
-        "server.lobbyconnection.config.FAF_POLICY_SERVER_BASE_URL",
-        f"http://{policy_server.host}:{policy_server.port}"
-    ):
+    mock_policy = mock.patch("server.lobbyconnection.config.FAF_POLICY_SERVER_BASE_URL",
+                    f"http://{policy_server.host}:{policy_server.port}")
```
Comment (Askaholic) [inline tests/integration_tests/conftest.py:111]:
```suggestion
    mock_policy = mock.patch(
        "server.lobbyconnection.config.FAF_POLICY_SERVER_BASE_URL",
        f"http://{policy_server.host}:{policy_server.port}"
    )
```

File: tests/integration_tests/test_login.py [hunk 29]
```diff
@@ -221,3 +224,135 @@ async def test_server_double_login(lobby_server):
...
+            "jti": "",
+            "client_id": ""
+        }, jwk_priv_key, algorithm="RS256", headers={"kid": jwk_kid}),
+        "unique_id": "some_id"
+    })
+
+    msg = await proto.read_message()
+    assert msg == {
+        "command": "authentication_failed",
+        "text": "User id not in database"
```
Comment (Askaholic) [inline tests/integration_tests/test_login.py:314]: need to update this test

**Review by Askaholic — COMMENTED** [review]
Running out of stuff to complain about so it must be pretty good :P

File: server/lobbyconnection.py (hunk 19 above)
Comment (Brutus5000) [reply server/lobbyconnection.py]: You mean something like send a token instead of a username/password?
I looked this up for unrealircd and anope but no matches. Since OAuth is very web-centric I guess OAuth is considered work of the devil for IRC power users :wink:

**Comment by Brutus5000** [comment]
If you need additional input from me, please ping me on Zulip. I have like 80 open notifications on Github.

**Comment by Sheikah45** [comment]
Fixes #440 just waiting on the user service to be implemented on prod

**Comment by Sheikah45** [comment]
The user service is being deployed on prod so this can now be considered mergeable.
@Askaholic can you fix the Pipfile?

**Comment by Askaholic** [comment]
Yea, I'll fix it.

File: server/oauth_service.py [hunk 30]
```diff
@@ -0,0 +1,52 @@
...
+    Service for managing the OAuth token logins and verification.
+    """
+
+    def __init__(self):
+        self.public_keys = {}
+
+    async def initialize(self) -> None:
+        await self.retrieve_public_keys()
+        # crontab: min hour day month day_of_week
+        # Run every day to update public keys.
```
Comment (Askaholic) [inline server/oauth_service.py:26]: Oh no, bad comment! This is why you should never write comments!

**Comment by Sheikah45** [comment]
Merging and making a release with this would have no impact on current operation correct?

File: server/lobbyconnection.py [hunk 31]
```diff
@@ -121,7 +126,15 @@ async def abort(self, logspam=""):
 
     async def ensure_authenticated(self, cmd):
         if not self._authenticated:
-            if cmd not in ["hello", "ask_session", "create_account", "ping", "pong", "Bottleneck"]:  # Bottleneck is sent by the game during reconnect
+            if cmd not in (
+                "ask_session",
+                "auth",
+                "Bottleneck"  # Bottleneck is sent by the game during reconnect
```
Comment (Askaholic) [inline server/lobbyconnection.py:132]: Oh wow, here's a nasty bug. What happens when you forget a comma:
```python3
>>> (
...     "ask_session",
...     "auth",
...     "Bottleneck"  # Bottleneck is sent by the game during reconnect
...     "create_account",
...     "hello",
...     "ping",
...     "pong",
... )
('ask_session', 'auth', 'Bottleneckcreate_account', 'hello', 'ping', 'pong')
```

File: tests/integration_tests/test_login.py [hunk 32]
```diff
@@ -221,3 +224,135 @@ async def test_server_double_login(lobby_server):
...
+        "user_agent": "faf-client",
+        "token": jwt.encode({
+          "sub": 3,
+          "user_name": "Rhiza",
+          "scope": [],
+          "exp": int(time() + 1000),
+          "authorities": [],
+          "non_locked": True,
+          "jti": "",
+          "client_id": ""
```
Comment (Askaholic) [inline tests/integration_tests/test_login.py:243]: Oof. only 2 space indent

File: server/lobbyconnection.py (hunk 31 above)
Comment (Sheikah45) [reply server/lobbyconnection.py:132]: can you even create an account through the server anymore? Didn't realize that was a command.

Comment (Askaholic) [reply server/lobbyconnection.py]: No, the command just returns an error message
